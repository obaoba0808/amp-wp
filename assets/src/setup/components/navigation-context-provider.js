/**
 * WordPress dependencies
 */
import { createContext, useState } from '@wordpress/element';
import { getQueryArg } from '@wordpress/url';

/**
 * External dependencies
 */
import PropTypes from 'prop-types';

export const Navigation = createContext();

/**
 * Context provider for navigating between and keeping track of pages in the app.
 *
 * @param {Object} props Component props.
 * @param {?any} props.children Component children.
 * @param {Array} props.pages Pages in the app.
 */
export function NavigationContextProvider( { children, pages } ) {
	// Initialize page from URL `amp-setup-screen` parameter. If not set, current page is 0.
	// This is primarily for testing.
	const [ activePageIndex, setActivePageIndex ] = useState( () => {
		const index = pages.findIndex( ( { slug } ) => slug === getQueryArg( global.location.href, 'amp-setup-screen' ) );
		return -1 < index ? index : 0;
	} );
	const [ canGoForward, setCanGoForward ] = useState( false );

	const currentPage = pages[ activePageIndex ];

	/**
	 * Navigates back to the previous page.
	 */
	const moveBack = () => {
		setActivePageIndex( activePageIndex - 1 );
		setCanGoForward( true );
	};

	/**
	 * Navigates to the next page. Pages are expected to set canGoForward to true when required actions have been taken.
	 */
	const moveForward = () => {
		setActivePageIndex( activePageIndex + 1 );
		setCanGoForward( false ); // Each page is responsible for setting this to true.
	};

	return (
		<Navigation.Provider
			value={
				{
					activePageIndex,
					canGoForward,
					currentPage,
					moveBack,
					moveForward,
					pages,
					setCanGoForward,
				}
			}
		>
			{ children }
		</Navigation.Provider>
	);
}

NavigationContextProvider.propTypes = {
	children: PropTypes.any,
	pages: PropTypes.arrayOf(
		PropTypes.shape( {
			PageComponent: PropTypes.func.isRequired,
			slug: PropTypes.string.isRequired,
			title: PropTypes.string.isRequired,
		} ),
	).isRequired,
};
