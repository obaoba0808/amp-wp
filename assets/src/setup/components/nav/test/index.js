
/**
 * External dependencies
 */
import { act } from 'react-dom/test-utils';
import { create } from 'react-test-renderer';

/**
 * WordPress dependencies
 */
import { render } from '@wordpress/element';

/**
 * Internal dependencies
 */
import { Nav } from '..';
import { NavigationContextProvider } from '../../navigation-context-provider';

let container;

const getNavButtons = ( containerElement ) => ( {
	nextButton: containerElement.querySelector( '.amp-setup-nav__next' ),
	prevButton: containerElement.querySelector( '.amp-setup-nav__prev' ),
} );

const MyPageComponent = () => <div />;
const testPages = [ { PageComponent: MyPageComponent, slug: 'slug', title: 'Page 0' }, { PageComponent: MyPageComponent, slug: 'slug-2', title: 'Page 1' } ];

describe( 'Nav', () => {
	beforeEach( () => {
		container = document.createElement( 'div' );
		document.body.appendChild( container );
	} );

	afterEach( () => {
		document.body.removeChild( container );
		container = null;
	} );

	it( 'matches snapshot', () => {
		const wrapper = create(
			<NavigationContextProvider pages={ testPages }>
				<Nav exitLink="http://site.test" />
			</NavigationContextProvider>,
		);
		expect( wrapper.toJSON() ).toMatchSnapshot();
	} );

	it( 'hides previous button on first page', () => {
		act( () => {
			render(
				<NavigationContextProvider pages={ testPages }>
					<Nav exitLink="http://site.test" />
				</NavigationContextProvider>,
				container,
			);
		} );

		const { nextButton, prevButton } = getNavButtons( container );

		expect( prevButton ).toBeNull();
		expect( nextButton ).not.toBeNull();
	} );

	it( 'disables next button on last page', () => {
		act( () => {
			render(
				<NavigationContextProvider pages={ [ testPages[ 0 ] ] }>
					<Nav exitLink="http://site.test" />
				</NavigationContextProvider>,
				container );
		} );

		const { nextButton } = getNavButtons( container );

		expect( nextButton.hasAttribute( 'disabled' ) ).toBe( true );
	} );
} );
