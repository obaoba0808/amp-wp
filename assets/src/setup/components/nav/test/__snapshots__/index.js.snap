// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Nav matches snapshot 1`] = `
<div
  className="amp-setup-nav"
>
  <div
    className="amp-setup-nav__close"
  >
    <a
      className="components-button is-link"
      href="http://site.test"
    >
      Close
    </a>
  </div>
  <div
    className="amp-setup-nav__prev-next"
  >
    <span
      className="amp-setup-nav__placeholder"
    >
       
    </span>
    <button
      className="components-button amp-setup-nav__next"
      disabled={true}
      onClick={[Function]}
      type="button"
    >
      <span
        className="amp-mobile-hide"
      >
        Next Step
      </span>
      <span
        className="amp-mobile-show"
      >
        Next
      </span>
      <svg
        aria-hidden={true}
        className="dashicon dashicons-arrow-right-alt2 amp-mobile-hide"
        focusable={false}
        height={18}
        role="img"
        viewBox="0 0 20 20"
        width={18}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M6 15l5-5-5-5 1-2 7 7-7 7z"
        />
      </svg>
    </button>
  </div>
</div>
`;
